README.md
# Synchronise #

This is the synchronise module.

### Overview ###

This module allows you to synchronise BitBucket Mercurial to GitHub Git 
projects.

### How do you make the synchonisation work? ###

There is an example directory.  Install it and follow these rules to 
synchronise your project called <project> on Bitbucket with user called 
<user> to Github:

1. Make sure your Bitbucket repository is Mercurial based. Your project 
   can be private or public.
2. In the Settings of your project (click the cogwheel), select the 
   Hooks section.  Select the 'POST' hook and point it to the URL you 
   installed this module in.  
   The URL needs to be public.  If that would be a problem, there is already 
   a synchroniser installed at 
   `http://api.elevenbits.org:8000/synchronise/`.  You can use that
   if you want to.
3. Make sure you have a GitHub project with the same username and the 
   same projectname, i.e. <user> and <project> respectively.
   *Note: Soon you'll be able to use different names for both users and 
   projects on the GitHub side.*
4. In the GitHub select your <project> and visit its Settings page.
   Add 'Synchroniser' as Collaborator.
5. Done!

From now on, when you do a push to your Bitbucket project, the module will 
be informed by the POST and push it to GitHub as the Synchroniser collaborator.
This will immediately make sure the GitHub project will be synchronised with
the Butbucket one.

### How does it work? ###

Each time a Mercurial push arrives at Bitbucket, it will send a request
to the URL specified in the POST hook of the pushed project.

The JSON content of the POST is clearly explained
[at the Bitbucket website](https://confluence.atlassian.com/display/BITBUCKET/POST+hook+management).

The '/synchronise/' link will parse the JSON and see if it is a valid POST.

When it is a valid POST it will clone the Mercurial repository specified in 
it using [hgapi](https://bitbucket.org/haard/hgapi). Then it will make
sure that the [hggit](https://bitbucket.org/durin42/hg-git) extension is 
enabled and an extra `github` path is added. After bookmarking the tip 
as master (via `hg bookmark master`), it will push the repo to GitHub (via
`hg push github`). Since Synchroniser is a collaborator the push will succeed.

See the code for more information.

### Error codes ###

Successful synchonisation request will return 200 with an informational
message.  The synchroniser will return 400 when a request was unsuccessfully
handled.

Some examples when a 400 would be returned:
- When invalid request arrives (e.g. no JSON payload)
- The github project specified in the request does not exist
- The user specified in the request the user does not exist
- No Synchroniser collaborator is specified in the GitHub project

A 500 will be returned when something bad happened at the server side, and
is not caused by an invalid request.  For example the hggit plugin is not
available or hgapi misbehaved.

### Can I create my own Synchroniser? ###

Yes you can! All the code is available here. When you have a server 
available, you can point the POST hook to your own machine.

Here is a first draft of the installation process using `nginx` and `uwsgi`:

* Make sure `nginx` and `uwsgi` are installed.
* Add a new site configuration in `/etc/nginx`.
* Add a new uwsgi configuration in `/etc/uwsgi`.
* Restart both `nginx` and `uwsgi`.
* Enjoy your synchronisation server!

#### Notes ####

* This is a Django project, and uses only Python based dependencies: hgapi
  and hggit.
* The database is not really needed.  The default sqlite3 will do for now.
* To install UWSGI:
   - Do not install using your distribution methods. So, do not use `apt-get`,
     `yum` or `pakman`.
   - Please download the latest uwsgi: http://projects.unbit.it/downloads/uwsgi-2.0.6.tar.gz
   - Build plugins for it:
     `python2.7 uwsgiconfig.py --plugin plugins/python core python27`
     `python3.4 uwsgiconfig.py --plugin plugins/python core python34`
   - Create two ini files, one for Python 2.7 (two.ini), another one for Python
     3 (three.ini):

##### two.ini #####  

[uwsgi]
plugin-dir = /home/jw/plugins
plugin = python27
chdir = /home/jw/apps/two
master = true
threads = 2
processes = 4
http-socket = :9090
wsgi-file = foobar.py

##### three.ini #####

[uwsgi]
plugin-dir = /home/jw/plugins
plugin = python34
chdir = /home/jw/apps/three
master = true
threads = 2
processes = 4
http-socket = :8080
wsgi-file = fizzbuzz.py

Start the uwsgi emperor with `uwsgi --emperor <path where the ini files live>`
and see the system handle to different python releases!

### Contribution guidelines ###

* Please write tests for me?
* Do not hesitate to do a code review.
* When writing code, do follow [PEP8](http://legacy.python.org/dev/peps/pep-0008/).
  But use `'` before `"`; so `print('Hello there')` and 
  `foo = '"Hej!", said Fritz'` are valid, while `print("Foobar")` is not. 

### Who do I talk to? ###

* Mail jw@elevenbits.com for info, help, support,...
* Leave requests and bug reports on the issues page


CHANGELOG

django-synchroniser 0.4.0 (released in three months)

    * Please add?

django-synchroniser 0.3.0 (released in three months)

    * Make sure the fabfile works properly.
    * Add git to git synchronisation?
    * Check task framework?

django-synchroniser 0.2.0 (released 09.12.2014)

    * Added git owner and git project vars in hook.
    * Logging properly now.

django-synchroniser 0.1.1 (released 21.07.2014)

    * Initial release.


synchronise/synchronise.py
from django.http import HttpResponse

import configparser
from os.path import join
from tempfile import TemporaryDirectory

import hgapi

import logging
logger = logging.getLogger(__name__)


def print_payload(payload, length):
    """
    Convert the payload to a string, and replace each whitespace section
    with a single space. Return it with a specified length.
    :param payload: A dict.
    :param length: The required length.
    :return: The processed payload of the specified length.
    """
    clean = ' '.join(str(payload).split())
    if len(clean) <= length:
        return clean
    else:
        return '"' + clean[0:length] + '..."'


def is_bitbucket_hg_payload(payload):
    """
    Check to see of the payload is a BitBucket POST from a Mercurial based
    project
    :param payload: The payload as a dist
    :return: True when it is a BitBucket POST from a Mercurial payload; False
    otherwise.
    """

    logger.debug('Checking {} for bitbucket and hg...'.
                 format(print_payload(payload, 20)))
    result = False
    try:
        if 'bitbucket' in payload['canon_url']:
            repo = payload['repository']
            if repo['scm'] == 'hg':
                logger.debug('This is a Bitbucket Mercurial repository POST.')
                result = True
    except KeyError:
        pass  # no canon_url or scm in payload
    return result


def bitbucket_hg_payload(payload, project=None, user=None):
    """
    Prepare a BitBucket Mercurial based project for synchronisation via
    HgGit
    :param payload: The payload as a dict
    :param project: The project string
    :param user: The userstring
    :return:
    """
    repo = payload['repository']
    # use default user and project when not given
    if user is None:
        user = repo['owner']
    if project is None:
        project = repo['name']
    # hg_path should be ssh://hg@bitbucket.org/<name>/<project>,
    hg_path = 'ssh://hg@bitbucket.org/{}/{}'.format(repo['owner'],
                                                    repo['name'])
    # git_path should be git+ssh://git@github.com/<user>/<project>.git
    git_path = 'git+ssh://git@github.com/{}/{}.git'.format(user, project)
    try:
        hg_to_git(hg_path, git_path)
    except hgapi.HgException as he:
        HttpResponse("Could not convert: {}.".format(he),
                     status=500, reason='HgApi problem: {}'.format(he))
    message = "Converted {} successfully".format(project)
    return HttpResponse(message, status=200, reason=message)


def synchronise(payload, user=None, project=None):
    """
    Synchronised the given payload.
    :param payload: The payload as a dict
    :param user: The user
    :param project: The project
    :return: A HttpResponse, either 200 when successful, 400 or 500 otherwise.
    """
    if is_bitbucket_hg_payload(payload):
        return bitbucket_hg_payload(payload, project, user)
    return HttpResponse('No proper JSON in payload. Discarding it.',
                        status=400, reason='No proper JSON in payload.')


def add_hggit_extension_and_git_path(hg_repo_path, git_path):
    """
    Add hggit extension to the hgrc and add git_path to its paths section.
    This is required to inform git of the hg changesets on the hg repository.
    :param hg_repo_path: The path of the hgrc
    :param git_path: The path of the github repo
    :return:
    """
    hgrc = join(hg_repo_path, '.hg', 'hgrc')
    config = configparser.ConfigParser()
    config.read(hgrc)
    # add hgit extension
    if 'extensions' in config.sections():
        if 'hggit' not in config['extensions']:
            config['extensions']['hggit']
    else:
        config['extensions'] = {}
        config['extensions']['hggit'] = ''
    # add git_path to the paths section
    if 'paths' in config.sections():
        if 'github' not in config['paths']:
            config['paths']['github'] = git_path
    else:
        config['paths'] = {}
        config['paths'][git_path] = ''
    # write the hgrc
    with open(hgrc, 'w') as config_file:
        config.write(config_file)


def hg_to_git(hg_path, git_path):
    """
    Convert the Bitbucket repo to a Github repo. First clone the hg_path,
    then add the hggit extension to this clone, and lastly push the clone
    to github.
    :param hg_path: The Bitbucket Mercurial based repository.
    :param git_path: The Github (obviously Git based) repository.
    :return:
    """
    logger.debug('Converting {} to {}.'.format(hg_path, git_path))
    with TemporaryDirectory(prefix='hg-') as hg_repo_path:
        logger.debug('Cloning {} to {}...'.format(hg_path, hg_repo_path))
        hg_repo = hgapi.hg_clone(hg_path, hg_repo_path)
        add_hggit_extension_and_git_path(hg_repo_path, git_path)
        logger.debug('Bookmarking the clone...')
        hg_repo.hg_bookmarks(action=hgapi.Repo.BOOKMARK_CREATE, name='master')
        logger.debug('Pushing it to github...')
        hg_repo.hg_push('github')


def git_to_hg(git_path, hg_path):
    logger.debug('Converting {} to {}.'.format(git_path, hg_path))
    raise NotImplemented


def git_to_git(git_path, other_git_path):
    logger.debug('Converting {} to {}.'.format(git_path, other_git_path))
    raise NotImplemented
